File: app/src/main/java/com/gps/gpsdome/service/GpsService.java
package com.gps.gpsdome.service;

import android.app.Service;
import android.content.Intent;
import android.os.IBinder;
import android.util.Log;

import androidx.annotation.Nullable;

/**
 * Created by xkai on 2019/10/30.
 */
public class GpsService extends Service {

    private Thread thread;
    private final static String TAG = "wzj";
    private int count;
    private boolean quit;

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    @Override
    public void onCreate() {
        super.onCreate();
        Log.d(TAG, "Service is invoke Created");
        thread = new Thread(new Runnable() {
            @Override
            public void run() {
                // 每间隔一秒count加1 ，直到quit为true。
                while (!quit) {
                    try {
                        Thread.sleep(5000);
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }
                    count++;
                    Log.d(TAG, "count:" + count);
                }
            }
        });
        thread.start();
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        return super.onStartCommand(intent, flags, startId);
    }

    @Override
    public void onDestroy() {
        Log.d(TAG, "Service is invoke Destroyed");
        Intent it = new Intent(getApplicationContext(), GpsService.class);
        startService(it);
        super.onDestroy();
    }
}

File: gps/src/main/java/com/xkai/gps/GpsService.kt
package com.xkai.gps

import android.app.Service
import android.content.Intent
import android.os.Handler
import android.os.IBinder
import android.os.Message
import android.os.Messenger
import android.os.RemoteException
import android.widget.Toast

//服务端service
class GpsService : Service() {
    companion object {
        private const val CODE = 1
        private const val TAG = "GpsService"
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        return START_STICKY
    }

    //创建一个送信人，封装handler
    private val mMessenger = Messenger(object : Handler() {
        override fun handleMessage(msg: Message) {
            val toClient = Message.obtain()
            when (msg.what) {
                CODE -> {
                    //接收来自客户端的消息，并作处理
                    val arg = msg.arg1
                    Toast.makeText(applicationContext, arg.toString() + "", Toast.LENGTH_SHORT)
                        .show()
                    toClient.arg1 = 1111111111
                    try {
                        //回复客户端消息
                        msg.replyTo.send(toClient)
                    } catch (e: RemoteException) {
                        e.printStackTrace()
                    }
                }
            }
            super.handleMessage(msg)
        }
    })

    override fun onBind(intent: Intent): IBinder? {
        return mMessenger.binder
    }
}
